sveltekit-authentication/index.md

```diff
@@ -0,0 +1,449 @@
+
+The JavaScript form action below will handle the actual request.
+The Directus SDK is not saving HttpOnly Cookies, thus the Directus API is directly accessed from SvelteKit in order to be able to save the tokens in a secure cookie via SvelteKit's cookie handler.
+
+```js [/signin/+page.server.js]
+import { redirect,fail } from '@sveltejs/kit';
+import { PUBLIC_APIURL } from '$env/static/public';
+import { constructCookieOpts } from '$lib/directus';
+
+const REFRESH_TOKEN_EXPIRATION = 30; // in days - adapt this with your Directus Settings
+
+export const load = async ({ locals,url }) => {
+	// redirect user if already logged in
+	if (locals.token) redirect(302, '/profile')
+	return {};
+}
+
+const loginUser = async (request,email,password) => {
+	let req = await fetch(`${PUBLIC_APIURL}/auth/login`, {
+		method: 'POST',
+		headers: {
+			'Content-Type': 'application/json',
+			'user-agent':request.headers.get("user-agent"),
+		},
+		body: JSON.stringify({
+			email,
+			password
+		})
+	});
+	if(req.status >= 300){
+		throw new Error(await req.text());
+	}
+	req = await req.json();
+	return req.data;
+}
+
+
+const login = async ({ cookies, request, url }) => {
+	const data = await request.formData();
+	const email = data.get('email');
+	const password = data.get('password');
+	const redirectedFrom = url.searchParams.get('redirectedFrom');
+	
+	try {
+		let tokens = await loginUser(request,email,password);
+
+		cookies.set('access_token',tokens.access_token, constructCookieOpts(Math.floor(tokens.expires/1000)));
+		cookies.set('refresh_token', tokens.refresh_token, constructCookieOpts(60 * 60 * 24 * REFRESH_TOKEN_EXPIRATION));
+	} catch (err) {
+		return fail(400, {email,message:err.message});
+	}
+
+	redirect(302, redirectedFrom ? redirectedFrom : `/profile`)
+}
+
+const register = async ({ cookies, request, url }) => {
+	const data = await request.formData();
+	const email = data.get('email');
+	const password = data.get('password');
+	const redirectedFrom = url.searchParams.get('redirectedFrom')
+	
+  //First create the user
+	let signupRequest = await fetch(`${PUBLIC_APIURL}/users`, {
+		method: 'POST',
+		headers: {
+			'Content-Type': 'application/json',
+			'user-agent':request.headers.get("user-agent")
+		},
+		body: JSON.stringify({
+			email,
+			password
+		})
+	});
+	if(signupRequest.status >= 300){
+		return fail(400, { email, message:await signupRequest.text() });
+	}
+	try {
+		let tokens = await loginUser(request,email,password);
+
+		cookies.set('access_token',tokens.access_token, constructCookieOpts(Math.floor(tokens.expires/1000)));
+		cookies.set('refresh_token', tokens.refresh_token, constructCookieOpts(60 * 60 * 24 * REFRESH_TOKEN_EXPIRATION));
+	} catch (err) {
+		return fail(400, err.message);
+	}
+	
+	redirect(302, redirectedFrom ? redirectedFrom : `/profile`)
+}
+
+export const actions = { login,register }
+```
+
```

Comment:
^ please break this down into smaller, explainable steps. This is a big snippet that contains too much functionality to just blindly copy and paste.

Reply:
i split it into two. more does not make sense, as it would create more overhead than help. i also added more comments.